routing/routing_integration_tests/bicycle_route_test.cpp

```diff
@@ -190,41 +228,92 @@ UNIT_TEST(Munich_OnewayBicycle2)
       mercator::FromLatLon(48.17867, 11.57303), 201.532 /* expectedRouteMeters */);
 }
 
-// https://github.com/organicmaps/organicmaps/issues/1603
 UNIT_TEST(London_GreenwichTunnel)
 {
-  // Has bicycle=yes, because it belongs to https://www.openstreetmap.org/relation/9063263
+  // Use the bicycle=dismount foot tunnel https://www.openstreetmap.org/way/4358990
+  // as a detour is way too long.
+  // https://github.com/organicmaps/organicmaps/issues/8028
 
-  /// @todo +50m detour is not suitable here, IMO.
   CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
       mercator::FromLatLon(51.4817397, -0.0100070258), {0.0, 0.0},
-      mercator::FromLatLon(51.4883739, -0.00809729298), 1305.81 /* expectedRouteMeters */);
+      mercator::FromLatLon(51.4883739, -0.00809729298), 1183.12 /* expectedRouteMeters */);
 }
 
 UNIT_TEST(Batumi_AvoidServiceDetour)
 {
+  // Go straight via a residential without a short detour via service road.
   CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
       mercator::FromLatLon(41.6380014, 41.6269446), {0.0, 0.0},
       mercator::FromLatLon(41.6392113, 41.6260084), 160.157 /* expectedRouteMeters */);
 }
 
 UNIT_TEST(Gdansk_AvoidLongCyclewayDetour)
 {
+  /// @todo(pastk): the preferred route goes by a shared cycleway also - replace with a better case (the next one!)
   CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
       mercator::FromLatLon(54.2632738, 18.6771661), {0.0, 0.0},
-      mercator::FromLatLon(54.2698882, 18.6765837), 775.81 /* expectedRouteMeters */);
+      mercator::FromLatLon(54.2698882, 18.6765837), 760.749 /* expectedRouteMeters */);
+}
+
+UNIT_TEST(Netherlands_AvoidLongCyclewayDetour)
+{
+  // Same as GraphHopper.
+  // The first sample from https://github.com/organicmaps/organicmaps/issues/1772
+  CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
+      mercator::FromLatLon(52.253405,6.182288), {0.0, 0.0},
+      mercator::FromLatLon(52.247599,6.197973), 1440.8 /* expectedRouteMeters */);
+}
+
+UNIT_TEST(Poland_AvoidDetourAndExtraCrossings)
+{
+  // Avoid making a longer cycleway detour which involves extra road crossings.
+  // Same as GraphHopper. Valhalla suggests going by the road (worse).
```

Comment:
I would say that https://www.openstreetmap.org/directions?engine=fossgis_valhalla_bicycle&route=50.082166%2C20.033060%3B50.082567%2C20.031853 is not really worse

I actually prefer it (and if you have no tolerance to car traffic you would not go into this road but detour via parks)

It is caused by fact that cycleway has two separate traffic lights - and you may be caught between two traffic lights between carriageways, while if you are in the middle of the crossing on road you can leave it
```suggestion
  // Same as GraphHopper. Valhalla suggests going by the road (worse for some cyclists, but changing light will not force cyclist to wait between two carriageways as they would be able to leave crossing).
```

Reply:
may be too much detail...
```suggestion
  // Same as GraphHopper. Valhalla suggests going by the road (worse for cyclists really preferring to avoid cars whenever possible, may be preferred for more aggressive ones depending on light phase to go through one traffic light only).
```